Stegos Chat

dbmcclain opened the pull request:
a work in progress

dbmcclain committed:
work in progress - first commit

adjust mesage payload to accommodate type byte indicators in serialize output size. Want total size = 2048 bytes

more work on Chat message validation
definition of Chat Groups & Channels & Owner Info
diving into serialization nightmare with Protobufs having thought I just conquered Serde... (sheesh!)

repartition responsibilities and share more code

add Double-Blind decryption routines

implement ECDH-style keying hints on recipient and owner
modify sender lookup
modify sender rekeying
modify cloaking for recipient, sender
modify signing of UTXOs by sender
modify message encrytion key derivation

adjust cargo.toml back to our chat needs including Dalek crate

modify expected UTXO size as a result of new key hinting

incremental progress - mostly Protobufs and Canaries

initial test probes on Chat Message UTXO size

perform Hash::rshift(-4) on chain derivation to bring all hash values into Fr domain
Gives us 97% chance of landing between (2^251, 2^252)

update validation tests for Chat

tying into the message pump

adorn code with GUI Alert

final estimate of where GUI needs to interact

Add Channel message processing

split Chat between blockchain and wallet

wow! - have to think like Fortran to make the code shorter

Fill in GUI API and some pseudo code for use at the bottom

Add subscribe_to_group() to also send initial keyings out to the group

Corrections to crypto - find_sender_newchain()
add API for subscribe_to_group()

Add logic in rekeying to search for prior epoch and transition to new owner_chain

Add infrastructure for recording chat group and channel messages that belong to owner, along with keying info, for later spending purposes. We record hash of UTXO, data of creation, and secret keying info needed to form Transaction signatures and proof of ownership. This needs to be capured at the time of arrival since chain codes are ephemeral.

initial carving for handling out of order delivery...

rtsisyk commented on blockchain/src/chat.rs:
```diff
@@ -0,0 +1,222 @@
+    /// Message sequence ID
+    pub sequence: u64,
+    // Segmentation info msg #x of #n
+    pub msg_nbr: u32,
```
I remember that we discussed that we don't want segmentation in the first version...

dbmcclain replied:
Yes we did. But that doesn't mean we will hold off forever... Just making allowance for future growth here.

dbmcclain replied:
There is one category of Chat messages that will produce benign fragmentation when you have more than 59 members in the group. It is helpful if all of the rekeying messages from one sender are identified with the same messsage ID, part M of N. Each fragment stands alone and does not need reassembly, but they are pieces of one interchange from the sender.

rtsisyk commented on blockchain/src/chat.rs:
```diff
@@ -0,0 +1,222 @@
+
+/// Chat Message UTXO.
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct ChatMessageOutput {
```
Why a new type of UTXO is needed? We inteionally increased the size of PaymentUTXO to 2k just to fit potential change messages!

dbmcclain replied:
There are things needed to be shown (e.g., sender pkey, creation date, message ID, signature) at the outer level of the Chat UTXO that have to be seen by validators and group members in order to validate and decrypt these special UTXO. The original payment UTXO was never designed with Chat in mind.

rtsisyk commented:
Any news here? Do we have anything for review?

dbmcclain committed:
members in rosters now keep public key, chain hash, and epoch of last update

remove subscribe_to_group(). On initial join, group owner has sent via private message, all membership info and our first chain code. No need to broadcast a flood of initial rekeying messages. Owner will inform the others in a group message what my public key and initial chain code are.

Add handling for owner generated NewMember messages carrying their (pkey, chain) pairs.

Add code for group owner to add list of new members to his own roster, and to send out NewMember messages to the group, containing their (public key, iniital chain) information. This is a group message, and hence encrypted.

dbmcclain commented:
To give a bit of background on the latest changes...

I have decided that, in order to avoid problems with reading out-of-order messages and stale messages, it is best if group owners initially set up a special chain code for use in all group rekeyings. 

That way, all new members will be informed of: owner pkey, owner chain, rekey chain, then list of all other group members showing public key, current chain of member, and epoch of last update seen by owner. This information is relayed through a private message from owner to new member.

Upon receipt of this information, the new member will construct a ChatSession struct, and add it to his Chat struct by calling subscribe_to_group(). 

Since the new member will not be in any other member's roster yet, the group owner must send a group message informing them all, of the new members public key and initial chain code. If we allow group owner to assign the first chain code for the new member, then we can avoid a rekeying storm from the new member.
 
The rekeying chain for the group never needs to change. If any member becomes evicted he will still be able to read all of the rekeying messages, but since at least one member - the most important one for crypto = group owner (who sent the eviction notice) - the primary member will not be sending a new chain code to the evicted member's public key. 

So even if all other members are still ignorant of the eviction, and they include the evicted member in their rekeyings, the evicted member will not know the new chain code of the group owner, to which all future group messages will be sent.

Hence, knowing the chain codes for all other members becomes useless since evicted member won't be able to find any group messages under the new rekeyings.

vldm commented:
@dbmcclain Lets, close this pullrequest for now and keep base branch for future development.

dbmcclain committed:
Merge pull request #1451 from stegos/dm-chat-members-with-epochs

chat members with epochs

dbmcclain committed:
Use deterministic randomness for new chain codes

rtsisyk committed:
Merge pull request #1453 from stegos/dm-chat-deterministic-randomness

Use deterministic randomness for new chain codes

dbmcclain committed:
make ALL uses of PRNG into deterministic randomness - this covers chain codes, and random cloaking factors on recipient and sender keys

vldm committed:
Merge pull request #1454 from stegos/dm-chat-all-randomness-deterministic

Make all Chat randomness deterministic

dbmcclain committed:
Corrections to use of detrand(). The chain code argument must be one which changes over time.

dbmcclain committed:
Add private messages needed for member join

recode byte dissection with LittleEndian::write_u32() - Thanks Vladimir!

dbmcclain committed:
Merge pull request #1455 from stegos/dm-corrections-to-detrand-use

Ensure ephemeral chain code argument to detrand()

dbmcclain committed:
Merge pull request #1458 from stegos/dm-chat-initial-private-messages

Dm chat initial private messages

dbmcclain committed:
Finish off Private Messaging for general users, add needed decoding logic to stegos_blockchain/src/output.rs

dbmcclain committed:
Merge pull request #1459 from stegos/dm-private-messsaging

Finish off Private Messaging for general users

dbmcclain committed:
make all chain codes represented as Fr not Hash

dbmcclain committed:
Merge pull request #1462 from stegos/dm-chains-to-Fr

All chain codes are Fr, not Hash

dbmcclain committed:
Add signatures to private message payloads for new-member messages

Add protective guards on decoding of encrypted vectors

dbmcclain committed:
Merge pull request #1465 from stegos/dm-chat-private-message-signatures

Add owner signature to private message payload for new-member messages
